### ISSUE TITLE
Configuration API  Building Block 

### ISSUE THREAD

**Issue by msfussell** [dapr/dapr#2941]
This is a building blocks proposal to include a configuration API for reading and writing application configuration data, for example from Azure Configuration Manager or GCP Configuration Management. 

https://docs.microsoft.com/en-us/azure/azure-app-configuration/overview and 
https://cloud.google.com/solutions/configuration-management/
https://aws.amazon.com/config/

By abstracting configuration API, this provide portability and pluggability of code. 

This could also be used for local and developer framework configurations abstractions , for example reading data from a local file or integration with ASP.NET Core configuration.

> /area runtime

## Release Note
RELEASE NOTE:

**Comment by jjcollinge** [comment]
Just an off the cuff thought. I've seen issues binding to cloud based config managers as when they become unavailable (due to an outage in AAD for example 👀) you cannot start your app. I wonder if Dapr could provide a fall-through mechanism to support falling back from a primary config provider to an alternative to help this scenario.

**Comment by KaiWalter** [comment]
Requirements I see currently (w/o checking Azure / AWS / GCP capabilities beforehand):

_API_
	
- GET to configuration scope - single Dapr service (=AppId) and an "application context" covering multiple semantically bounded services
- GET single configuration value
- GET configuration structure (collection of all values) for a scope (see above)
- PUSH (like pub/sub) when single configuration value is changed

_sidecar_

- once configuration values are pushed, those are cached and returned also in cache the underlying service is not available
- cache gets invalidated updated when configuration values change in underlying service

_component definition_

- be able to set defaults for configuration values in cache underlying configuration service is not available


Generally, if shaped right this could even used for feature switching.

**Comment by KaiWalter** [comment]
suggestion for the API surface

# for PULL of configuration values

to pull single value
`GET http://localhost:<daprPort>/v1.1/configuration/<configuration-context>/<key>`

to pull multiple values
`POST/PUT http://localhost:<daprPort>/v1.1/configuration/<configuration-context>/bulk`
```
{
          "keys": [ "key1", "key2" ]
}
```

# for PUSH of configuration values

basically add something similar to pub/sub

general to provide configuration routes
`GET http://localhost:<appPort>/dapr/configuration`
```
[
  {
    "context": "configuration-context",
    "route": "/configuration-change-bulk"
  },
  {
    "context": "configuration-context",
    "key": "key1",
    "route": "/configuration-change-key1"
  }
]
```

**or in .NET**
for bulk
[ConfigurationTopic("configuration-context")]
public async Task OnMultiConfigurationChange(IDictionary<string, string> kv)
{
...
}

for single value
[ConfigurationTopic("configuration-context", "key1")]
public async Task OnSingleConfigurationChange(string key, string value)
{
...
}

**Comment by msfussell** [comment]
> Just an off the cuff thought. I've seen issues binding to cloud based config managers as when they become unavailable (due to an outage in AAD for example 👀) you cannot start your app at all. I wonder if Dapr could provide a fall-through mechanism to support falling back from a primary config provider to an alternative to help this scenario for those who deem it necessary.

Yes, this is a real issue when using centralized configuration stores and something that we have experienced in Azure. What you do not want is the app inability to start due to a configuration service being unavailable

**Comment by withinboredom** [comment]
> once configuration values are pushed, those are cached and returned also in cache the underlying service is not available

Where will these be cached at? Do we need a cache building block or will Dapr act as a distributed cache?

**Comment by KaiWalter** [comment]
implying that configuration values generally should not consume much space/memory - I'd like to have these cached as close as possible : in Dapr sidecar

**Comment by skyao** [comment]
@msfussell Hi, Mark, I have finished the design of configuration api and submit a issue here

https://github.com/dapr/dapr/issues/2988

**Comment by rhummelmose** [comment]
> > once configuration values are pushed, those are cached and returned also in cache the underlying service is not available
> 
> Where will these be cached at? Do we need a cache building block or will Dapr act as a distributed cache?

Very much agree these would have to be cached in the sidecar
